### PR TITLE
Replace gulp-responsive with sharp

### PR DESCRIPTION
**Pull request by nick-mon1**
## Summary
Replace `gulp-responsive` with `sharp` for generating responsive image sizes addresses:

- image darkening when uploading images
- reduces responsive image files from **33** to **4**
- saves 75% disk space on average per upload (most variants are not used)

#### Image Darkening

Updates to `sharp` in [december](https://github.com/GSA/digitalgov.gov/commit/7fccbeed0fbe11a73f07dc1f8188988174aad418#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519) moved from `31.3` to `32.6` which caused image darkening. `gulp-responsive` uses `sharp` under the hood and prevented updating due to lack of maintenance.

<details><summary>Before</summary>
<p>

<img width="364" alt="Screenshot 2024-04-09 at 9 53 38 AM" src="https://github.com/GSA/digitalgov.gov/assets/104778659/bb0739b6-ea27-4810-b8a9-cdfe9b9e3f35">

</p>
</details> 

<details><summary>After</summary>
<p>

<img width="347" alt="Screenshot 2024-04-09 at 9 53 45 AM" src="https://github.com/GSA/digitalgov.gov/assets/104778659/235c17be-8e15-4c07-93fc-3b54cc397af8">


</p>
</details> 


#### Unused Variants
33 variants are created per image upload, most are unused.

**before**

| size     | variant   | ext  | count |
| -------- | --------- | ---- | ----- |
| original |           | jpg  | 1     |
| original |           | webp | 1     |
| original | bu (blur) | jpg  | 1     |
| original | bw        | jpg  | 1     |
| original | bw        | webp | 1     |
| 200      |           | jpg  | 1     |
| 200      |           | webp | 1     |
| 200      | bw        | jpg  | 1     |
| 200      | bw        | webp | 1     |
| 400      |           |      | 4     |
| 600      |           |      | 4     |
| 800      |           |      | 4     |
| 1200     |           |      | 4     |
| 1600     |           |      | 4     |
| 2400     |           |      | 4     |

**after**

| size     | variant  |
| -------- | -------- |
| original | jpg\|png |
| 200      | jpg\|png |
| 400      | jpg\|png |
| 800      | jpg\|png |

#### Disk Space

Reducing the number of variants can save us 75% disk space per upload going forward.
A future enhancement will add `webp` format.

| # images | disk space |
| ------ | ----- |
| 33     | 2.4M  |
| 4      | 557KB |


### Solution

#### Refactor Javascript
There are 3 javascript files that are run sequentially to generate the responsive images.
`gulp-responsive` was used in `file-process.js` and is now replaced with `sharp` (55c7dc0b396b1184377bc2f7d551895df08d8955).

**config/gulp/**
| file              | function                                                                 |
| ----------------- | ------------------------------------------------------------------------ |
| `file-prep.js`    | cleans filename, checks if image file, create `yml` data file            |
| `file-process.js` | generates responsive image variants with `sharp` and working directories |
| `file-upload.js`  | uploads responsive variants to S3 and removes working directories        |


#### Shortcodes

`img.html` and `img-right.html` partials have been updated to only display these below sizes (441b01495afdd96c795c7d08f525d154d6b504bb). 

| size     | variant  |
| -------- | -------- |
| original | jpg\|png |
| 200      | jpg\|png |
| 400      | jpg\|png |
| 800      | jpg\|png |


#### Follow Up Work

- add `webp` format, further saving more disk space
- remove existing unused image variants on S3


### How To Test

<details><summary>Test Images</summary>
<p>

![accessibility-for-teams-guide](https://github.com/GSA/digitalgov.gov/assets/104778659/0775de6c-b6a9-476b-9f43-1aed89745f65)
![cartoon-artwork](https://github.com/GSA/digitalgov.gov/assets/104778659/f71bcbd1-f7ae-41d9-9992-b578143409ee)
![USWDS Monthly Call Feb 2024](https://github.com/GSA/digitalgov.gov/assets/104778659/122fca1f-9829-46ee-9760-a13ee96d03b5)


</p>
</details> 

1. run locally `npx gulp upload` with no images in `content/upload/_inbox`folder should display "no images in folder" in cli
2. run `npx gulp upload` with image(s) in the `content/upload/_inbox` will process 4 variants for each image (original, _w200, _w400, _w800)
3. You will see new variants saved to `content/upload/_inbox/working-images/processed`

**Regression Test**
Test that each new variant (200, 400, 800) is served.


| shortcode | page                                                                                                                                                                                                             |
| --------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| img       | [_w200](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/nl-replace-gulp-with-sharp/2020/01/28/improving-experience-federal-grant-reporting/)                                                                    |
| img       | [_w400](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/nl-replace-gulp-with-sharp/resources/an-introduction-to-working-with-social-media-influencers/) |
| img       | [_w800](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/nl-replace-gulp-with-sharp/2023/12/19/amplifying-customer-voices/)                              |
| img-right | [_w400](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/nl-replace-gulp-with-sharp/resources/an-introduction-to-working-with-social-media-influencers/) |


Serve `_w800` instead of `_w1200` image.

| before                                                                                           | after                                                                                                                                                                                                           |
| ------------------------------------------------------------------------------------------------ | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| [serve 1200px](https://digital.gov/2023/12/08/making-gsa-public-art-collection-more-accessible/) | [serve 800px](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/nl-replace-gulp-with-sharp/2023/12/08/making-gsa-public-art-collection-more-accessible/) |

### What To Test

- 4 images are generated
- filename reads: original-image-name_w200.png
- regression test pages serve only original, _w200, _w400, and _w800 images

#### Testing Notes

The images will not be uploaded due to running locally and S3 credentials will display the following error:
<details><summary>S3 Error</summary>
<p>

```bash
[09:47:51] 'uploadImage' errored after 5.2 s
[09:47:51] Error in plugin "gulp-s3-upload"
Message:
    S3 headObject Error: Error: connect EHOSTDOWN 169.254.169.254:80 - Local (10.0.71.0:58060)
    at internalConnect (node:net:1090:16)
    at defaultTriggerAsyncIdScope (node:internal/async_hooks:464:18)
    at node:net:1315:9
    at process.processTicksAndRejections (node:internal/process/task_queues:77:11)
Details:
    domainEmitter: [object Object]
    domainThrown: false
```
</p>
</details>

When you run `npx gulp upload` the `working-images` directory will be created but not deleted due to the deletion command not executing because of the S3 error that will occur.

<details><summary>Image directories</summary>
<p>

<img width="412" alt="Screenshot 2024-04-09 at 9 48 01 AM" src="https://github.com/GSA/digitalgov.gov/assets/104778659/39506a0f-d3d5-48d5-91d1-4a466fa0e92b">

</p>
</details> 


---

### Dev Checklist

- [x] PR has correct labels
- [x] A11y testing (voice over testing, meets WCAG, run axe tools)
- [x] Code is formatted properly

### CODE REVIEW

**Comment by clmedders** [comment]
Screenshot 2024-04-10 at 1 05 47 PM

@nick-mon1 im getting this error in my terminal when i run the `nps gulp upload` after i have added the test images...

**Review by clmedders — APPROVED** [review]

**Review by mejiaj — CHANGES_REQUESTED** [review]
@nick-mon1, in terminal could you run `npx eslint ./config/gulp/file-process.js` and fix the low LOE alerts? We can follow up on the rest in the next sprint.

**Comment by nick-mon1** [comment]
@mejiaj 

Refactored code to remove all errors. 
Running `npx eslint ./config/gulp/file-process.js` outputs:

```bash

/Users/nlyons/Documents/gsa/digitalgov.gov/config/gulp/file-process.js
   66:9   warning  Unexpected console statement  no-console
   84:11  warning  Unexpected console statement  no-console
  101:7   warning  Unexpected console statement  no-console
  115:7   warning  Unexpected console statement  no-console
  124:7   warning  Unexpected console statement  no-console
  133:3   warning  Unexpected console statement  no-console

✖ 6 problems (0 errors, 6 warnings)
```

**Review by clmedders — APPROVED** [review]
LGTM

I tested with the npx script and i am only get a warning but those console statments are used for debugging in the event of an error

File: config/gulp/file-process.js [hunk 1]
```diff
@@ -1,280 +1,137 @@
...
+    image
+  );
+
+  const imageVariant = {
+    name: imageName,
+    extension: imageExtension,
+    path: imagePath,
+  };
+  return imageVariant;
+}
```
Comment (mejiaj) [inline config/gulp/file-process.js:52]: ## Optional

You could leverage [Path.parse()](https://nodejs.org/api/path.html#pathparsepath) here to use the native method for getting this information.

**Review by mejiaj — COMMENTED** [review]
Code looks much better overall, thank you. Added some feedback based on a few test cases.

### 1. No images

Running `npx gulp upload` with an empty `content/upload/_inbox` directory I see the following message:

```bash
npx gulp upload
[09:26:23] Using gulpfile ~/web/digitalgov.gov/gulpfile.js
[09:26:23] Starting 'upload'...
[09:26:23] Starting 'fileTidy'...
[09:26:23] Finished 'fileTidy' after 376 μs
[09:26:23] Starting 'cleanInbox'...
(node:47441) NOTE: We are formalizing our plans to enter AWS SDK for JavaScript (v2) into maintenance mode in 2023.

Please migrate your code to use AWS SDK for JavaScript (v3).
For more information, check the migration guide at https://a.co/7PzMCcy
(Use `node --trace-warnings ...` to show where the warning was created)
[09:26:23] Finished 'cleanInbox' after 4.49 ms
[09:26:23] Starting 'writeDataFile'...
[09:26:23] Finished 'writeDataFile' after 5.35 ms
[09:26:23] Starting 'processImages'...
[09:26:23] Finished 'processImages' after 199 μs
[09:26:23] Starting 'removeProcessedImage'...
Removing processed images
No images to process
Error processing file: [undefined]. ENOENT: no such file or directory, scandir './content/uploads/_working-images/to-process'
[09:26:23] Finished 'removeProcessedImage' after 1.09 ms
[09:26:23] Starting 'uploadImage'...
starting image upload
[09:26:23] Finished 'uploadImage' after 7.52 ms
[09:26:23] Starting 'uploadFile'...
starting file upload
[09:26:23] Finished 'uploadFile' after 1.15 ms
[09:26:23] Starting 'cleanup'...
[09:26:23] Finished 'cleanup' after 784 μs
[09:26:23] Finished 'upload' after 23 ms
```

This _looks_ to start three tasks: `upload, fileTidy, and cleanInbox`.  

We could improve clarity here by first checking the directory and **only** running the other tasks if there's a file.


```mermaid
graph TD;
    images_empty-->Yes;
    images_empty-->No;
    Yes-->Return;
    No-->processImages;
    processImages-->uploadImage;
    processImages-->writeDataFile;
    uploadImage-->cleanup;
```

Also curious, why do `uploadImage` and `uploadFile` run at the same time.

**Comment by nick-mon1** [comment]
@mejiaj 

These gulp `tasks` are all run with the `series` command, they will run no matter the condition of the previous task. The [API docs](https://gulpjs.com/docs/en/api/concepts) do not provide a way to exit a task on a condition gracefully. We could throw an error but this will cause build errors with **circleci**:

```javascript
process.exit();
```

This produces:

```bash
[16:37:40] Using gulpfile ~/Documents/gsa/digitalgov.gov/gulpfile.js
[16:37:40] Starting 'upload'...
[16:37:40] Starting 'fileTidy'...
[16:37:40] Finished 'fileTidy' after 663 μs
[16:37:40] Starting 'cleanInbox'...
(node:30491) NOTE: We are formalizing our plans to enter AWS SDK for JavaScript (v2) into maintenance mode in 2023.

Please migrate your code to use AWS SDK for JavaScript (v3).
For more information, check the migration guide at https://a.co/7PzMCcy
(Use `node --trace-warnings ...` to show where the warning was created)
[16:37:40] Finished 'cleanInbox' after 9.22 ms
[16:37:40] Starting 'writeDataFile'...
[16:37:40] Finished 'writeDataFile' after 11 ms
[16:37:40] Starting 'processImages'...
[16:37:40] Finished 'processImages' after 311 μs
[16:37:40] Starting 'removeProcessedImage'...
Removing processed images
No images to process
[16:37:40] The following tasks did not complete: upload, <series>, removeProcessedImage
[16:37:40] Did you forget to signal async completion?
```

We would need to refactor these tasks across 3 files to add better job control and error handling, which looks like limited for our needs based on the `gulp` API.

An alternative would be to remove the reliance on `gulp` and use **es modules** to handoff each of these functions (prep, process, upload) that would allow us to have more control of the process.

**Comment by clmedders** [comment]
is it smart to get rid of gulp?  since USWDS uses it or maybe since this is outside of that package then we might be ok?  I see what James is saying and I thought that we would always have some kind of image there so it wouldn't be an issue.  Just some thoughts...

**Review by mejiaj — APPROVED** [review]
Approving because this issue is to fix image processing & composition is expanding scope.

You can use conditionals and have quite a bit of freedom composing tasks. See GulpJS docs on [composing tasks](https://gulpjs.com/docs/en/getting-started/creating-tasks/#compose-tasks).

**Review by clmedders — APPROVED** [review]

**Pull request merged**